### PR TITLE
Optimize 2x2 blocks by relying on spdm

### LINKED ISSUE #524: Optimizing gemm usage in splines
In `SplinesLinearProblem2x2Blocks::solve()`, two `gemm` operations are performed, involving bottom-left and top-right corners. Those are stored as dense matrix (2D Kokkos::View). However, the bottom-left corner contains a very small number of non-zeros (here 6 for a `4x96` block):

image

 while the top-right corner has the following structure (once factorized as `Q^-1*gamma`):
 
image

Of course, dense storages for those blocks and dense gemm usage are suboptimal. We could consider improving this using conversion to sparse format (removing zeros) during factorization and using some `spmv`.

@yasahi-hpc

### PR DESCRIPTION
**Pull request by blegouix**
Closes #524 

For 100 iterations of 1000x100000 `characteristics_advection`.
image

Before the current solution I tried using different `spmv` implementations from KokkosKernels and even re-implemented a small `KOKKOS_FUNCTION SerialSpmv` but the mandatory CSR storage really seems unappropriate for the non-zeros distribution of our cases, thus the performance was not satisfying. So I just wrote a simple `spdm` (sparse-dense mutiplication) based on Coo, and it works pretty well.

### CODE REVIEW

File: CMakeLists.txt
Comment (blegouix) [inline CMakeLists.txt:268]: Maybe this is overkill to ask user to compile all KokkosSparse just to get the CooMatrix class. I can use a custom struct if you want.

**Comment by blegouix** [comment]
As #480 is stalled but virtually approved, I think this can be reviewed. We can merge before of after we merge #480 in the main (to have 1 or 2 squashes), as you prefer.

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:186]:
```suggestion
                                rows_idx(n_nonzeros.d_view()) = i;
                                cols_idx(n_nonzeros.d_view()) = j;
                                values(n_nonzeros.d_view()) = aij;
                                n_nonzeros.d_view()++;
```

Let's avoid too complicated lines

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:183]:
```suggestion
                            if (Kokkos::abs(aij) >= 1e-14) {
```

I think we can strictly compare to 0 ? The zeros come non-modifications of the matrix, not from the result of a computation ?

**Review by tpadioleau — CHANGES_REQUESTED** [review]
I find it interesting. I have one question though that you may not like, could we use Ginkgo for these kernels ?

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (blegouix) [reply include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:45]: So, in the first place I was trying to use KokkosKernels but at the end I realized the only thing I needed was the KokkosSparse::MatrixCoo class which requires to compile the whole KOKKOS_KERNELS_SPARSE component, which seems overkill, so I redevelopped my own Coo implementation.

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (blegouix) [reply include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:183]: There is a computation for the top-right corner block during factorization, which is Q^-1*gamma, which look like this:
image

**Comment by blegouix** [comment]
About the Ginkgo question, of course we can store the COO in a Ginkgo object. About the operation `spdm`, I can check, but would we want to rely on Ginkgo in a class which is more for the LAPACK backend ? Especially for something so simple (no solver) ?

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (blegouix) [reply include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:45]: Cf https://github.com/CExA-project/ddc/pull/526/commits/7a48e7ceba910e7ec3c294c1c40da919f081feeb which can be reverted

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (blegouix) [inline include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:270]:
```suggestion
     * @param LinOp The sparse matrix, left side of the matrix multiplication.
     * @param x The dense matrix, right side of the matrix multiplication.
     * @param y The dense matrix to be altered by the operation.
```

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (tpadioleau) [reply include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:185]: I see, do we loose a lot of performance if we only remove exact zeroes ?

Comment (blegouix) [reply include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:185]: Yes (I guess a lot of almost-zero are retained):

image

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (tpadioleau) [reply include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:45]: If there is no equivalent function of `spdm_minus1_1` it might not be interesting indeed.

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:53]:
```suggestion
        Coo()
            : nrows(0)
            , ncols(0)
        {
        }
```

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:270]:
```suggestion
     * @param[in] LinOp The sparse matrix, left side of the matrix multiplication.
     * @param[in] The dense matrix, right side of the matrix multiplication.
     * @param[inout] y The dense matrix to be altered by the operation.
```

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:160]:
```suggestion
    Coo dense2coo(Kokkos::View<const double**, Kokkos::LayoutRight, typename ExecSpace::memory_space>
```

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:156]:
```suggestion
     * @param[in] dense_matrix The dense storage matrix whose non-zeros are extracted to fill the COO matrix.
```

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:185]: Please make this threshold an input parameter of `dense2coo`.

**Review by tpadioleau — CHANGES_REQUESTED** [review]

File: include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp
Comment (blegouix) [reply include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp:45]: Yes, `spmv` is an equivalent on the paper but it does not work like mine (COO+parallelized along m-rhs, sequential loop on the elements of the COO) and has no satisfying performance for our case.

**Comment by tpadioleau** [comment]
Can you rebase on main ?

**Comment by blegouix** [comment]
There are a lot of conflicts. Can we just wait for #480 to be merged ?

**Comment by tpadioleau** [comment]
> There are a lot of conflicts. Can we just wait for #480 to be merged ?

Yes I missed the fact it was based on branch `kokkos-kernels-backend`

**Review by tpadioleau — APPROVED** [review]
Seems good to me!

**Pull request merged**
